File: src/mediaScanner.ts
import type { MediaKind, ScannedMedia } from './types';

/**
 * Injected into the WebView on demand. Finds <video>/<source> elements, .gif <img> tags
 * and direct links to video/gif files on the currently loaded page, then reports them
 * back via `window.ReactNativeWebView.postMessage`.
 *
 * `blob:` and `data:` URIs are intentionally skipped: they cannot be re-downloaded with a
 * plain HTTP request (e.g. MSE-based streaming players such as YouTube commonly only
 * expose a `blob:` URL for the active video).
 */
export const SCAN_JS = `
(function () {
  var urls = [];
  var seen = {};
  function add(src, kind) {
    if (!src) return;
    var abs;
    try { abs = new URL(src, document.baseURI).href; } catch (e) { return; }
    if (abs.indexOf('blob:') === 0 || abs.indexOf('data:') === 0) return;
    if (seen[abs]) return;
    seen[abs] = true;
    urls.push({ url: abs, kind: kind });
  }
  var videoExt = /\\.(mp4|webm|mov|m4v|mkv|3gp|avi)(\\?|#|$)/i;
  var gifExt = /\\.gif(\\?|#|$)/i;
  document.querySelectorAll('video').forEach(function (v) {
    add(v.currentSrc, 'video');
    add(v.getAttribute('src'), 'video');
    v.querySelectorAll('source').forEach(function (s) {
      add(s.getAttribute('src'), 'video');
    });
  });
  document.querySelectorAll('img').forEach(function (img) {
    var s = img.currentSrc || img.getAttribute('src') || '';
    if (gifExt.test(s)) add(s, 'gif');
  });
  document.querySelectorAll('a[href]').forEach(function (a) {
    var href = a.getAttribute('href') || '';
    if (videoExt.test(href)) add(href, 'video');
    else if (gifExt.test(href)) add(href, 'gif');
  });
  window.ReactNativeWebView.postMessage(JSON.stringify({ type: 'mediaScanResult', items: urls }));
  true;
})();
`;

export function parseScanMessage(raw: string): ScannedMedia[] | null {
  let data: unknown;
  try {
    data = JSON.parse(raw);
  } catch {
    return null;
  }
  if (
    typeof data !== 'object' ||
    data === null ||
    (data as { type?: unknown }).type !== 'mediaScanResult' ||
    !Array.isArray((data as { items?: unknown }).items)
  ) {
    return null;
  }
  const items = (data as { items: unknown[] }).items;
  const result: ScannedMedia[] = [];
  for (const entry of items) {
    if (
      typeof entry === 'object' &&
      entry !== null &&
      typeof (entry as any).url === 'string' &&
      ((entry as any).kind === 'video' || (entry as any).kind === 'gif')
    ) {
      result.push({ url: (entry as any).url, kind: (entry as any).kind });
    }
  }
  return result;
}

/** Builds unique, filesystem-safe file names for a freshly scanned list of media. */
export function buildFileNames(scanned: ScannedMedia[]): string[] {
  const usedNames = new Map<string, number>();
  return scanned.map((media, index) => {
    let name = deriveFileName(media.url, media.kind, index + 1);
    const occurrence = usedNames.get(name) ?? 0;
    usedNames.set(name, occurrence + 1);
    if (occurrence > 0) {
      const dot = name.lastIndexOf('.');
      name = dot > 0 ? `${name.slice(0, dot)}_${occurrence}${name.slice(dot)}` : `${name}_${occurrence}`;
    }
    return name;
  });
}

function deriveFileName(url: string, kind: MediaKind, fallbackIndex: number): string {
  const defaultExt = kind === 'gif' ? 'gif' : 'mp4';
  let name = '';
  try {
    const { pathname } = new URL(url);
    const lastSegment = pathname.split('/').filter(Boolean).pop();
    if (lastSegment) name = decodeURIComponent(lastSegment);
  } catch {
    // Malformed URL - fall back to the generated name below.
  }
  if (!name) {
    name = `media_${fallbackIndex}.${defaultExt}`;
  } else if (!name.includes('.')) {
    name = `${name}.${defaultExt}`;
  }
  return name.replace(/[\\/:*?"<>|]/g, '_');
}

export function formatFileSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  const kb = bytes / 1024;
  if (kb < 1024) return `${kb.toFixed(0)} KB`;
  const mb = kb / 1024;
  if (mb < 1024) return `${mb.toFixed(1)} MB`;
  const gb = mb / 1024;
  return `${gb.toFixed(2)} GB`;
}

File: app/src/main/java/de/carstenkeller/videodownloader/MediaScanner.kt
package de.carstenkeller.videodownloader

import android.net.Uri
import org.json.JSONArray
import org.json.JSONTokener

data class ScannedMedia(val url: String, val kind: MediaKind, val posterUrl: String? = null)

/**
 * Finds <video>/<source> elements, .gif <img> tags and direct links to video/gif files
 * in the currently loaded page - including same-origin <iframe> content, recursively.
 * Blob: and data: URIs are intentionally skipped since they cannot be re-downloaded with
 * a plain HTTP request (e.g. MSE-based streaming players such as YouTube commonly only
 * expose a blob: URL). Cross-origin iframes cannot be inspected either - the browser's
 * same-origin policy blocks script access to their content, with no workaround from JS.
 */
object MediaScanner {

    const val SCAN_JS = """
        (function() {
          var urls = [];
          var seen = {};
          function absolutize(url, doc) {
            if (!url) return null;
            try { return new URL(url, doc.baseURI).href; } catch (e) { return null; }
          }
          function add(src, kind, poster, doc) {
            var abs = absolutize(src, doc);
            if (!abs) return;
            if (abs.indexOf('blob:') === 0 || abs.indexOf('data:') === 0) return;
            if (seen[abs]) return;
            seen[abs] = true;
            urls.push({ url: abs, kind: kind, poster: absolutize(poster, doc) });
          }
          var videoExt = /\.(mp4|webm|mov|m4v|mkv|3gp|avi)(\?|#|${'$'})/i;
          var gifExt = /\.gif(\?|#|${'$'})/i;
          function scanDocument(doc) {
            try {
              doc.querySelectorAll('video').forEach(function(v) {
                var poster = v.getAttribute('poster');
                add(v.currentSrc, 'video', poster, doc);
                add(v.getAttribute('src'), 'video', poster, doc);
                v.querySelectorAll('source').forEach(function(s) {
                  add(s.getAttribute('src'), 'video', poster, doc);
                });
              });
              doc.querySelectorAll('img').forEach(function(img) {
                var s = img.currentSrc || img.getAttribute('src') || '';
                if (gifExt.test(s)) add(s, 'gif', null, doc);
              });
              doc.querySelectorAll('a[href]').forEach(function(a) {
                var href = a.getAttribute('href') || '';
                if (videoExt.test(href)) add(href, 'video', null, doc);
                else if (gifExt.test(href)) add(href, 'gif', null, doc);
              });
              doc.querySelectorAll('iframe').forEach(function(frame) {
                try {
                  var innerDoc = frame.contentDocument;
                  if (innerDoc) scanDocument(innerDoc);
                } catch (e) {
                  // Cross-origin iframe: blocked by the browser's same-origin policy.
                }
              });
            } catch (e) {
              // Defensive: one broken frame should not abort the whole scan.
            }
          }
          scanDocument(document);
          return JSON.stringify(urls);
        })();
    """

    fun parseResult(rawEvaluateResult: String?): List<ScannedMedia> {
        if (rawEvaluateResult == null || rawEvaluateResult == "null") return emptyList()
        val jsonString = try {
            JSONTokener(rawEvaluateResult).nextValue() as? String
        } catch (e: Exception) {
            null
        } ?: return emptyList()

        val array = try {
            JSONArray(jsonString)
        } catch (e: Exception) {
            return emptyList()
        }

        val result = mutableListOf<ScannedMedia>()
        for (i in 0 until array.length()) {
            val obj = array.optJSONObject(i) ?: continue
            val url = obj.optString("url").takeIf { it.isNotBlank() } ?: continue
            val kind = if (obj.optString("kind") == "gif") MediaKind.GIF else MediaKind.VIDEO
            val poster = obj.optString("poster").takeIf { it.isNotBlank() && it != "null" }
            result.add(ScannedMedia(url, kind, poster))
        }
        return result
    }

    /**
     * Builds unique, filesystem-safe file names for a freshly scanned list of media.
     * [alreadyUsedNames] lets repeated scans (e.g. after scrolling for more lazy-loaded
     * content) avoid colliding with files already shown from an earlier scan.
     */
    fun buildFileNames(scanned: List<ScannedMedia>, alreadyUsedNames: Set<String> = emptySet()): List<String> {
        val usedNames = mutableMapOf<String, Int>()
        alreadyUsedNames.forEach { usedNames[it] = 1 }
        return scanned.mapIndexed { index, media ->
            var name = deriveFileName(media.url, media.kind, index + 1)
            val occurrence = usedNames.getOrDefault(name, 0)
            usedNames[name] = occurrence + 1
            if (occurrence > 0) {
                val dot = name.lastIndexOf('.')
                name = if (dot > 0) {
                    "${name.substring(0, dot)}_$occurrence${name.substring(dot)}"
                } else {
                    "${name}_$occurrence"
                }
                usedNames[name] = 1
            }
            name
        }
    }

    private fun deriveFileName(url: String, kind: MediaKind, fallbackIndex: Int): String {
        val defaultExt = if (kind == MediaKind.GIF) "gif" else "mp4"
        val lastSegment = try {
            Uri.parse(url).lastPathSegment?.substringBefore('?')
        } catch (e: Exception) {
            null
        }
        var name = lastSegment?.let { runCatching { Uri.decode(it) }.getOrDefault(it) }.orEmpty()
        if (name.isBlank()) {
            name = "media_$fallbackIndex.$defaultExt"
        } else if (!name.contains('.')) {
            name = "$name.$defaultExt"
        }
        return name.replace(Regex("[\\\\/:*?\"<>|]"), "_")
    }
}

fun formatFileSize(bytes: Long): String {
    if (bytes < 1024) return "$bytes B"
    val kb = bytes / 1024.0
    if (kb < 1024) return String.format("%.0f KB", kb)
    val mb = kb / 1024.0
    if (mb < 1024) return String.format("%.1f MB", mb)
    val gb = mb / 1024.0
    return String.format("%.2f GB", gb)
}
